hare/broker.go

```diff
@@ -327,124 +267,61 @@ func (b *Broker) HandleEligibility(ctx context.Context, em *types.HareEligibilit
 	return false
 }
 
-func (b *Broker) getInbox(id types.LayerID) chan any {
-	b.mu.RLock()
-	defer b.mu.RUnlock()
-	out, exist := b.outbox[id.Uint32()]
-	if !exist {
-		return nil
-	}
-	return out
-}
-
-func (b *Broker) handleEarlyMessage(logger log.Log, layer types.LayerID, nodeID types.NodeID, msg any) error {
+// Register a layer to receive messages
+// Note: the registering instance is assumed to be started and accepting messages.
+func (b *Broker) Register(ctx context.Context, id types.LayerID) (chan any,
+	error,
+) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	layerNum := layer.Uint32()
-	if _, exist := b.pending[layerNum]; !exist { // create buffer if first msg
-		b.pending[layerNum] = make([]any, 0, inboxCapacity)
+	if !id.After(b.latestLayer) { // should expect to update only newer layers
+		b.WithContext(ctx).With().Error("tried to update a previous layer",
+			log.Stringer("this_layer", id),
+			log.Stringer("prev_layer", b.latestLayer))
 	}
 
-	// we want to write all buffered messages to a chan with InboxCapacity len
-	// hence, we limit the buffer for pending messages
-	if len(b.pending[layerNum]) == inboxCapacity {
-		logger.With().Warning("too many pending messages, ignoring message",
-			log.Int("inbox_capacity", inboxCapacity),
-			log.Stringer("smesher", nodeID))
-		return nil
+	// We check here since early messages could have already constructed this map entry
+	outboxCh, exist := b.outbox[id.Uint32()]
+	if !exist {
+		outboxCh = make(chan any, inboxCapacity)
+		b.outbox[id.Uint32()] = outboxCh
 	}
-	b.pending[layerNum] = append(b.pending[layerNum], msg)
-	return nil
-}
 
-func (b *Broker) cleanOldLayers() {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	for i := b.minDeleted.Add(1); i.Before(b.latestLayer); i = i.Add(1) {
-		_, exist := b.outbox[i.Uint32()]
-
-		if !exist { // unregistered
-			b.minDeleted = b.minDeleted.Add(1)
-		} else { // encountered first still running layer
-			break
+	oversize := len(b.outbox) - b.limit
+	if oversize >= 0 {
+		// Remove earliest layers to make space for this layer
+		k := sortedKeys(b.outbox)
+		for i := 0; i < oversize; i++ {
+			delete(b.outbox, k[i])
+			b.With().Info("unregistered layer due to maximum concurrent processes limit", types.NewLayerID(k[i]))
 		}
 	}
-}
 
-// Register a layer to receive messages
-// Note: the registering instance is assumed to be started and accepting messages.
-func (b *Broker) Register(ctx context.Context, id types.LayerID) (chan any, error) {
-	b.setLatestLayer(ctx, id)
+	b.latestLayer = id
 
-	// check to see if the node is still synced
-	if !b.Synced(ctx, id) {
```

Comment:
i think this is where i mentioned redundant during our offline discussion. not the gossip handler path.

Reply:
Sorry, I'm not understanding. What are you suggesting?

Reply:
when we discussed offline, i did mention that checking for syncedness here is redundant. i thought that is why you removed the syncedness check everywhere

Reply:
So we now only check syncedness when handling messages.